### ISSUE TITLE
Weight Normalization on Pretrained Models?

### ISSUE THREAD

**Issue by flashbob** [pytorch/fairseq#839]
Hey everyone, thanks for the great library!
I have a question regarding the pretrained models you provide for NMT.

I'm planning on using the pretrained nmt models as base for some transfer learning experiments. Are the downloadable model files already weight normalized?

I figured out that if I use pytorch.hub to load the model, fairseq seems to apply WN with the method make_generation_faster, and making the model not trainable. 

But if I comment out the section where make_generation_faster is applied during loading of the model, I get a model that does not complain when I execute model.train(). So I now assume, that the models were saved BEFORE weight normalization was applied and they could be used for further training or fine tuning??
Just want to make sure that I didn't miss anything, so any feedback would be appreciated!

In case I'm right, it would be nice to have a flag that gives the option to set weight normalization to True/False when loading a model from pytorch.hub! :)

**Comment by myleott** [comment]
> So I now assume, that the models were saved BEFORE weight normalization was applied and they could be used for further training or fine tuning??

Correct. We save checkpoints so that they can be used to resume training/finetuning. At inference time we then call make_generation_fast, which can make changes to the model in order to improve inference speed, but which may break finetuning.

> In case I'm right, it would be nice to have a flag that gives the option to set weight normalization to True/False when loading a model from pytorch.hub!

Good idea, we will provide this.

**Comment by vumaasha** [comment]
@myleott I am interested in sending a PR for this change. @flashbob Could you help with some additional details /steps to reproduce what you were trying to do

**Comment by flashbob** [comment]
@vumaasha I want to use torch.hub to load a pretrained model and avoid having fairseq apply weight normalization to it so I can resume training with some other data for transfer learning.

The Generator class in hub_utils.py does not take an argument that makes weight normalization optional:
`# optimize model for generation
        for model in self.models:
            model.make_generation_fast_(
                beamable_mm_beam_size=(
                    None if getattr(args, 'no_beamable_mm', False)
                    else getattr(args, 'beam', 5)
                ),
                need_attn=getattr(args, 'print_alignment', False),
            )`

So I just commented this out to load the model and it worked (I guess :D).
Would be as simple as adding an additional flag "make_generation_fast = True/false" to give as a parameter to the torch.hub command and placing the code above in an "if make_generation_fast" statement.
